_posts/2024-04-10-benchmarks-over-time.md

```diff
@@ -0,0 +1,547 @@
+
+```sql
+SELECT 
+    sum(v3) as v3 
+FROM <dataframe or parquet file>
+```
+
+### Window Functions
+
+We also added an entire series of Window function benchmarks. 
+Window functions are a critical workload in real world data analysis scenarios, and can stress test a system in other ways.
+DuckDB has implemented state of the art algorithms to quickly process even the most complex window functions.
+We use the largest table from the join benchmark as the raw data for these new tests to help with comparability to the rest of the benchmark. 
+
+Window function benchmarks are much less common than more traditional joins and aggregations, and we were unable to find a suitable suite off the shelf. 
+These queries were designed to showcase the variety of uses for window functions, but there are certainly more that could be added.
+We are open to your suggestions for queries to add, and hope these queries could prove useful for other systems!
+
+Since the window functions benchmark is new, the window functions from each of the queries included are shown in the appendix at the end of the post.
+
+### Workload Size
+
+We test only the middle 5GB dataset size for the workloads mentioned thus far, primarily because some import and export operations to external formats like Pandas must fit in memory (and we used a Macbook Pro M1 with only 16GB of RAM). 
+Additionally, running the tests for 3 year's of DuckDB versions was time-intensive even at that scale, due to the performance of older versions.
+
+### Scale Tests
+
+However, using only 5GB of data does not answer our second key question: "What scale of data can it handle?"!
+We also ran only the group by and join related operations (avoiding in-memory imports and exports) at the 5GB and the 50GB scale. 
+Older versions of DuckDB could not handle the 50GB dataset when joining or aggregating, but modern versions can handle both, even on a memory-constrained 16GB RAM laptop.
+Instead of measuring performance, we measure the size of the benchmark that was able to complete on a given version.
+
+### Summary Metrics
+
+With the exception of the scale tests, each benchmark was run 3 times and the median time was used for reporting results.
+The scale tests were run once and produced a binary metric, success or failure, at each data size tested. 
+As older versions would not fail gracefully, the scale metrics were accumulated across multiple partial runs.
+
+### Computing Resources
+
+All tests use a Macbook Pro M1 with 16GB of RAM.
+In 2024, this is far from state of the art! 
+If you have more powerful hardware, you will see both improved performance and scalability.
+
+### DuckDB Versions
+
+Version 0.2.7 was the first version to include a Python client compiled for ARM64, so it was the first version that could easily run on the benchmarking compute resources. 
+Version 1.0.0 is the latest available at the time of publication, although we also provide a sneak preview of an in-development feature branch.
```

Comment:
At the time I had run it, I was pointed at feature! Things move pretty fast around here...haha. 

It would be maybe an hour of work to re-run the analysis on main and re-output all the charts if that would be better. (it would take ~5 hours of runtime if we redo the scale one, so that would have to wait to run overnight).

Reply:
Keeping feature due to timing!